runtime/vm/xz64/UpcallThunkGen.cpp

```diff
@@ -0,0 +1,922 @@
+					gprIndex += 1;
+				}
+				break;
+			case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_SP_DP:
+			case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_SP_SP_DP:
+			case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_DP_SP:
+			case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_DP_SP_SP:
+			case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_MISC_SP:
+			case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_MISC_DP:
+			case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_SP_MISC:
+			case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_DP_MISC:
+			case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_OTHER:
+			case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_MISC:
+				// There are two possible scenarios here. Either the struct is passed by value in a GPR (if struct size is 1,2,4,8),
+				// or by a pointer in GPR. In either case, we just need to store whatever is in the GPR into memory.
+				// So we do that here.
+				if (gprIndex <= maxGPRIndex) {
+					thunkMem += STG(thunkMem, gprIndex, 0, R15, offsetToParamArea + (stackSlotIndex * 8));
+					stackSlotIndex += 1;
+				}
+				gprIndex += 1;
+				break;
+			default:
+				Assert_VM_unreachable();
+		}
+	}
+
+	// Now jump to the common dispatcher helper routine.
+	// For the first parameter materialize the metaData pointer into R2.
+	// For the second parameter, load C_SP + 160 into R3 to point to where the current list of parameters start.
+	thunkMem += IIHF(thunkMem, R2, (reinterpret_cast<uint64_t>(metaData) >> 32));
+	thunkMem += IILF(thunkMem, R2, reinterpret_cast<uint64_t>(metaData));
+	// Load the R3 with argPointer (stackFrame + 160)
+	thunkMem += LAY(thunkMem, R3, 0, R15, offsetToParamArea);
+	// Materialize the destination address in R1.
+	thunkMem += IIHF(thunkMem, R1, reinterpret_cast<uint64_t>(metaData->upCallCommonDispatcher) >> 32);
+	thunkMem += IILF(thunkMem, R1, reinterpret_cast<uint64_t>(metaData->upCallCommonDispatcher));
+	// Now generate the branch to the helper
+	thunkMem += BASR(thunkMem, R14, R1);
+
+	// Upon returning copy back the structure, then restore the stack pointer.
+	if (hiddenParameter) {
+		// If there's a hidden parameter, then we need to copy the result back into the buffer provided by the native caller.
+		// The max displacement value possible by an MVC instruction is 4096 bytes. So if the struct size is smaller than 4096,
+		// we can copy all of it in a straight sequence of MVC instructions. If it's larger, we fall through into the `else` block
+		// below and copy the data in a loop.
+		if (sigArray[lastSigIdx].sizeInByte <= maxMVCDisplacement)
+			thunkMem += generateSmallStructCopyInstructions(thunkMem, (I_64)sigArray[lastSigIdx].sizeInByte, offsetToParamArea);
+		else
+			thunkMem += generateLargeStructCopyInstructions(thunkMem, (I_64)sigArray[lastSigIdx].sizeInByte, offsetToParamArea);
+	}
+	thunkMem += LAY(thunkMem, R15, 0, R15, frameSize);
+	// Now load R14 with the return address.
+	thunkMem += LG(thunkMem, R14, 0, R15, offsetToR14Slot);
+	// Then branch back
+	thunkMem += BASR(thunkMem, R14, R14);
```

Comment:
We do not return after this point right? In this case, we should use `BCR`.

Reply:
This should be fixed in: c54d5bd7f1575e9af8aa7f5c2fd37c396ce8be05